sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/PullOutJoinConditionSuite.scala

```diff
@@ -0,0 +1,133 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.optimizer
+
+import org.apache.spark.sql.catalyst.dsl.expressions._
+import org.apache.spark.sql.catalyst.dsl.plans._
+import org.apache.spark.sql.catalyst.expressions.{Alias, Coalesce, IsNull, Literal, Substring, Upper}
+import org.apache.spark.sql.catalyst.plans._
+import org.apache.spark.sql.catalyst.plans.logical._
+import org.apache.spark.sql.catalyst.rules._
+import org.apache.spark.sql.internal.SQLConf
+
+class PullOutJoinConditionSuite extends PlanTest {
+
+  private object Optimize extends RuleExecutor[LogicalPlan] {
+    val batches =
+      Batch("Pull out join condition", Once,
+        PullOutJoinCondition,
+        CollapseProject) :: Nil
+  }
+
+  private val testRelation = LocalRelation('a.string, 'b.int, 'c.int)
+  private val testRelation1 = LocalRelation('d.string, 'e.int)
+  private val x = testRelation.subquery('x)
+  private val y = testRelation1.subquery('y)
+
+  test("Pull out join keys evaluation(String expressions)") {
+    val joinType = Inner
+    Seq(Upper("y.d".attr), Substring("y.d".attr, 1, 5)).foreach { udf =>
+      val originalQuery = x.join(y, joinType, Option("x.a".attr === udf))
+        .select("x.a".attr, "y.e".attr)
+      val correctAnswer = x.select("x.a".attr, "x.b".attr, "x.c".attr)
+        .join(y.select("y.d".attr, "y.e".attr, Alias(udf, udf.sql)()),
+          joinType, Option("x.a".attr === s"`${udf.sql}`".attr)).select("x.a".attr, "y.e".attr)
+
+      comparePlans(Optimize.execute(originalQuery.analyze), correctAnswer.analyze)
+    }
+  }
+
+  test("Pull out join keys evaluation(null expressions)") {
+    val joinType = Inner
+    val udf = Coalesce(Seq("x.b".attr, "x.c".attr))
```

Comment:
Can't we merge this test case with the one above?

Reply:
`Seq(Upper("d".attr), Substring("d".attr, 1, 5), Coalesce(Seq("b".attr, "c".attr)))...`

Reply:
Move this test to `SQLQuerySuite` to test infer more `IsNotNull`: https://github.com/apache/spark/blob/4aebe0a19fdf84516e30ab917eb149d39c4d4ac6/sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala#L4231-L4253